test/core/channel/channel_tracer_test.cc

```diff
@@ -0,0 +1,217 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/useful.h>
+
+#include "src/core/lib/channel/channel_tracer.h"
+#include "src/core/lib/iomgr/exec_ctx.h"
+
+#include "test/core/util/channel_tracing_utils.h"
+#include "test/core/util/test_config.h"
+
+using grpc_core::ChannelTracer;
+
+static void add_simple_trace(ChannelTracer* tracer) {
+  tracer->AddTrace(grpc_slice_from_static_string("simple trace"),
+                   GRPC_ERROR_CREATE_FROM_STATIC_STRING("Error"),
+                   GRPC_CHANNEL_READY, nullptr);
+}
+
+// checks for the existence of all the required members of the tracer.
+static void validate_tracer(ChannelTracer* tracer,
+                            size_t expected_num_nodes_logged,
+                            size_t max_nodes) {
+  if (!max_nodes) return;
+  char* json_str = tracer->RenderTrace(true);
+  grpc_json* json = grpc_json_parse_string(json_str);
+  validate_channel_data(json, expected_num_nodes_logged,
+                        GPR_MIN(expected_num_nodes_logged, max_nodes));
+  grpc_json_destroy(json);
+  gpr_free(json_str);
+}
+
+static void validate_tracer_data_matches_uuid_lookup(ChannelTracer* tracer) {
+  intptr_t uuid = tracer->GetUuid();
+  if (uuid == -1) return;  // Doesn't make sense to lookup if tracing disabled
+  char* tracer_json_str = tracer->RenderTrace(true);
+  char* uuid_lookup_json_str =
+      ChannelTracer::GetChannelTraceFromUuid(uuid, true);
+  strcmp(tracer_json_str, uuid_lookup_json_str);
+  gpr_free(tracer_json_str);
+  gpr_free(uuid_lookup_json_str);
+}
+
+// ensures the tracer has the correct number of children tracers.
+static void validate_children(ChannelTracer* tracer,
+                              size_t expected_num_children) {
+  char* json_str = tracer->RenderTrace(true);
+  grpc_json* json = grpc_json_parse_string(json_str);
+  validate_json_array_size(json, "children", expected_num_children);
+  grpc_json_destroy(json);
+  gpr_free(json_str);
+}
+
+static void test_basic_channel_tracing(size_t max_nodes) {
```

Comment:
a simple one liner saying what exactly is being tested maybe?

Reply:
Done